lib/core/test/unit/Cardano/Wallet/Api/ServerSpec.hs

```diff
@@ -265,3 +324,140 @@ errorHandlingSpec = describe "liftHandler and toServerError" $ do
+
+    in
+        checkCoverage $ cover 35 (isNonTrivial res) "non-trivial" $
+            keys _ours `disjoint` keys _foreign
+  where
+    labeledCheck notF f a b = counterexample
+        (show a <> " " <> notF <> " " <> show b)
+        (property $ f a b)
+
+    disjoint = labeledCheck "is not disjoint from" Set.disjoint
+
+prop_listStakeKeysBalance
+    :: PrettyShow UTxO
+    -> SaneOurStakeKeysInput
+    -> Property
+prop_listStakeKeysBalance (PrettyShow utxo) (SaneOurStakeKeysInput ours) =
+    let
+        res@ApiStakeKeys{_ours, _foreign, _none } = runIdentity $
+            listStakeKeys' utxo accountOfAddress (const $ pure mempty) ours
+
+    in
+        checkCoverage $ cover 35 (isNonTrivial res) "non-trivial" $
+            mconcat
+                [ totalStake _ours
+                , totalStake _foreign
+                , totalStake [_none]
+                ]
+                ===
+            TokenBundle.getCoin (balance utxo)
+
+-- | Type for where the reward accounts and indices are unique.
+--
+-- This is required for some properties, and does not seem worthwhile to protect
+-- against.
+newtype SaneOurStakeKeysInput
+    = SaneOurStakeKeysInput [(RewardAccount, Natural, ApiWalletDelegation)]
+    deriving (Show, Eq)
+
+mockOurStakeKeysInput :: Set RewardAccount -> SaneOurStakeKeysInput
+mockOurStakeKeysInput oursSet = SaneOurStakeKeysInput
+    $ zipWith (\ix acc -> (acc, ix, noDelegation))
+        [(0::Natural) ..]
+        (Set.toList oursSet)
+
+instance Arbitrary SaneOurStakeKeysInput where
+    arbitrary = mockOurStakeKeysInput <$> arbitrary
+    shrink (SaneOurStakeKeysInput x) = map mockOurStakeKeysInput $
+        shrink (Set.fromList (map (\(acc,_,_) -> acc) x))
+
+isNonTrivial :: ApiStakeKeys (n :: NetworkDiscriminant) -> Bool
+isNonTrivial ApiStakeKeys{_ours,_foreign,_none} =
+    length (keys _ours)> 1
+    && length (keys _foreign) > 1
+    && totalStake [_none] > Coin 0
+
+-- | Get the set of @RewardAccount@s from a list of stake keys (works with both
+-- ours and foregin).
+keys
+    :: (HasField' "_key" a (ApiT RewardAccount, proxy))
+    => [a]
+    -> Set RewardAccount
+keys = Set.fromList . map (getApiT . fst . view #_key)
+
+-- | Get the total stake of a list of stake keys (works with ours-, foreign-, or
+-- none-keys).
+totalStake
+    :: (HasField' "_stake" a (Quantity "lovelace" Natural))
+    => [a]
+    -> Coin
+totalStake = foldMap (Coin . fromIntegral . getQuantity . view #_stake)
+
+-- | Just something to pass listStakeKeys
+noDelegation :: ApiWalletDelegation
+noDelegation = ApiWalletDelegation
+    (ApiWalletDelegationNext NotDelegating Nothing Nothing)
+    []
+
+accountOfAddress :: Address -> Maybe RewardAccount
+accountOfAddress (Address bytes) =
+    let
+        str = B8.unpack bytes
+        char = last str
+    in
+        -- Leveraging the fact that addresses from genUTxOIndexSmall are
+        -- "ADDR{0 .. 7}"
+        if char == '7'
```

Comment:
If you're looking for something less fragile, then the address→account mapping could be stored in the generated test case.